### PR TITLE
Parallel dimension map

### LINKED ISSUE #1052: Unsafe omission of predicates when number of threads is larger than IterDomain extent
https://github.com/csarofeen/pytorch/blob/5778648a2fb9b7394a02dd1af9e3c1f8891e5f9a/torch/csrc/jit/codegen/cuda/index_compute.cpp#L2256-L2262

### LINKED ISSUE #1053: Predication bug with non-unique threading dimension
This fusion fails at the validation:

```
TEST(NVFuserTest, FusionNonUniqueThreadDim_CUDA) {
  auto fusion = std::make_unique<Fusion>();
  FusionGuard fg(fusion.get());

  auto tv0 = makeSymbolicTensor(1);
  fusion->addInput(tv0);
  auto tv1 = sum(tv0, {0});
  fusion->addOutput(tv1);

  auto tv2 = add(tv0, new Double(1));
  fusion->addOutput(tv2);

  //tv1->split(0, 32);
  tv1->split(0, 8);
  auto tv1_rf = tv1->rFactor({-1});

  tv1_rf->computeAt(tv1, 1);

  tv1_rf->axis(-1)->parallelize(ParallelType::TIDx);

  tv2->axis(0)->parallelize(ParallelType::TIDx);

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  at::Tensor input1 = at::randn({32}, options);

  auto at_tv1 = (input1).sum({0});
  auto at_tv2 = input1 + 1;

  FusionExecutor fe;
  fe.compileFusion(fusion.get());
  auto outputs = fe.runFusion({input1});
  testValidate(
      fusion.get(), outputs, {input1}, {at_tv1, at_tv2}, __LINE__, __FILE__);
}
```

The generated code is:

```
__global__ void kernel1(Tensor<float, 1> T0, Tensor<float, 0> T1, Tensor<float, 1> T2) {
  alignas(4) extern __shared__ char array[];
  void* shared_mem = array;
  T2[(((nvfuser_index_t)threadIdx.x) * 1)]
    = T0[(((nvfuser_index_t)threadIdx.x) * T0.stride[0])]
    + (float) 1;
  T1[0] = 0;
  #pragma unroll 1
  for(nvfuser_index_t ki45 = 0; ki45 < (ceilDiv(T0.size[0], 8)); ++ki45) {
    float T3[1];
    T3[0] = 0;
    blockReduce<true, false, false>(
      T3[0],
      T0[(((ki45 * 8) + ((nvfuser_index_t)threadIdx.x)) * T0.stride[0])],
      [](float &a, float b) { a = a + b; },
      threadIdx,
      blockDim,
      static_cast<float*>(shared_mem),
      (((ki45 * 8) + ((nvfuser_index_t)threadIdx.x)) < T0.size[0]),
      float(0));
    if (((((nvfuser_index_t)threadIdx.x) == 0))) {
      T1[0]
        = T1[0]
        + T3[0];
    }
  }
}
```

The problem is the predication for `T3` blockReduce. `TIDx` is mapped to `T3.axis(1)` and `T2.axis(0)`, which are 8 and 32, respectively, so `blockDim.x` is 32. Since there are 32 threads, predcating the `T3` block reduction with `(((ki45 * 8) + ((nvfuser_index_t)threadIdx.x)) < T0.size[0])` is not correct. I think it should also need: `threadIdx.x < 8`. But doing so always seems to increase the predication overhead significantly.  

This problem also happens with padding for warp reduction. See https://github.com/csarofeen/pytorch/pull/976#issuecomment-897917911.

### PR DESCRIPTION
**Pull request by naoyam**
This is a follow-up to #1054.

Adds `ParallelDimensionMap` as yet another temporary map during lowering. It maps for each of `TIDx/y/z` and `BIDx/y/z` to its dimension. By default, it's mapped to `blockDim.x/y/z` and `gridDim.x/y/z`, respectively, but a single unique constant dimension is found, that constant value is used instead. The map is used to determine IterDomain extents in indexing.

With this, in a simple case, like `tv->split(0, 128)` and `tv->axis(1)->parallelize(ParallelType::TIDx)`, the generated code would still use `128` as the extent of the second axis. That was the case before, but changed after #1054 to use `blockDim.x`. 

If not constant extent is found, the map returns `blockDim` and `gridDim`. So, for example, when a domain with extent `ceilDiv(ki100, 128)`  is mapped to `TIDx`, it would appear as `blockDim.x`, whereas previously it was `ceilDiv(ki100, 128)`. Both are correct, but I think just using `blockDim` or `gridDim` could be more efficient. 

The map also provides a query function to see if a ParallelType has a dimension that exactly matches with the extents of all mapped domains. This is intended to be used to fix #1052, though I haven't tried it yet.

Fixes #1053.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+      } else {
+        // Unmatched. This dom_set extents may not be unique.
+        all_equal = false;
+        break;
+      }
+    }
+
+    // At this point, it still remains undetermined whether this id
+    // matches with those previously looked at. Constant check failed,
+    // but symbolic matching may succeed.
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp:167]: Note to myself that both constant and symbolic checks need to handle padding.

File: torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+  if (dim1 == dim2) {
+    return true;
+  }
+
+  // When Both are Int, they are same if both have the same constant
+  auto dim1_int = dynamic_cast<kir::Int*>(dim1);
+  auto dim2_int = dynamic_cast<kir::Int*>(dim2);
+  if (dim1_int && dim2_int) {
+    if (dim1_int->isConst() && dim2_int->isConst()) {
+      return dim1_int->value() == dim2_int->value();
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp:222]: nitpick: may need an expression evaluator for constant expressions, as `constCheck` will be recursive.

File: torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+  TORCH_INTERNAL_ASSERT(isParallelTypeThread(pt), "Invalid ParallelType: ", pt);
+  auto it = dim_map_.find(pt);
+  if (it == dim_map_.end()) {
+    return nullptr;
+  } else {
+    return it->second;
+  }
+}
+
+bool ParallelDimensionMap::isExact(ParallelType pt) const {
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp:198]: Is this method used for testing only or there will be other places that will rely on it?

**Review by shmsong — APPROVED** [review]
Looks good to me! 

Had some minor questions and comments.

File: torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp:198]: I intend to use it to fix #1052. Haven't tried it yet.

File: torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp:222]: `isConst` here doesn't use `constCheck`.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp:222]: Oh I see. Thanks.

**Pull request merged**